File: app/servicos/reconciliacao.py
"""Fase 4 - os cinco relatorios de reconciliacao da migracao.

  1. orfaos de processo    - cartoes/linhas sem NUP
  2. orfaos de parecer     - pareceres sem processo, e processos que deveriam
                             ter parecer e nao tem
  3. conflito de numeracao - o mesmo numero apontando para coisas diferentes
  4. NUPs com DV invalido  - digitados sem conferencia
  5. divergencia de marco  - marco inicial que nao bate com a portaria

Nenhum deles resolve nada sozinho: expoem para o Fabricio decidir.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.modelos import (
    MigracaoRejeitada,
    ParecerTecnico,
    Processo,
    StgPlanilhaParecer,
    StgTrelloCartao,
)
from app.servicos import datas_br, nup as servico_nup
from app.servicos.importacao_trello import parecer_candidato


@dataclass
class Achado:
    referencia: str
    detalhe: str
    processo_id: int | None = None
    parecer_id: int | None = None


@dataclass
class Reconciliacao:
    orfaos_processo: list[Achado] = field(default_factory=list)
    orfaos_parecer: list[Achado] = field(default_factory=list)
    conflitos_numeracao: list[Achado] = field(default_factory=list)
    nups_dv_invalido: list[Achado] = field(default_factory=list)
    divergencias_marco: list[Achado] = field(default_factory=list)

    @property
    def total(self) -> int:
        return sum(
            len(x)
            for x in (
                self.orfaos_processo,
                self.orfaos_parecer,
                self.conflitos_numeracao,
                self.nups_dv_invalido,
                self.divergencias_marco,
            )
        )

    def como_secoes(self) -> list[tuple[str, str, list[Achado]]]:
        return [
            (
                "Órfãos de processo",
                "cartões e linhas sem NUP — precisam do número real para virar processo",
                self.orfaos_processo,
            ),
            (
                "Órfãos de parecer",
                "pareceres sem processo vinculado, e o contrário",
                self.orfaos_parecer,
            ),
            (
                "Conflito de numeração",
                "o mesmo número apontando para coisas diferentes — não resolver "
                "automaticamente, decidir caso a caso",
                self.conflitos_numeracao,
            ),
            (
                "NUPs com dígito verificador inválido",
                "entraram com a flag de dispensa; confirmar no SEI",
                self.nups_dv_invalido,
            ),
            (
                "Divergência de marco inicial",
                "a data do marco não bate com a publicação da portaria (RN-05)",
                self.divergencias_marco,
            ),
        ]


def _orfaos_processo(s: Session) -> list[Achado]:
    achados: list[Achado] = []
    for cartao in s.execute(select(StgTrelloCartao)).scalars():
        if not servico_nup.extrair(cartao.descricao):
            achados.append(
                Achado(
                    referencia=f"trello:{cartao.card_id}",
                    detalhe=f"cartão '{cartao.nome}' na lista '{cartao.lista}' sem NUP",
                )
            )
    for linha in s.execute(select(StgPlanilhaParecer)).scalars():
        if linha.col_b_numero_parecer and not linha.col_k_numero_processo:
            achados.append(
                Achado(
                    referencia=f"{linha.arquivo}:{linha.linha_origem}",
                    detalhe=(
                        f"parecer {linha.col_b_numero_parecer}/{linha.col_d_ano or '?'} "
                        "sem nº de processo na planilha"
                    ),
                )
            )
    # NUPs sinteticos criados na migracao do Trello (ano 1900)
    for processo in s.execute(
        select(Processo).where(Processo.nup.like("23086.%/1900-%"))
    ).scalars():
        achados.append(
            Achado(
                referencia=processo.nup,
                detalhe="NUP sintético gerado na migração — substituir pelo número real",
                processo_id=processo.id,
            )
        )
    return achados


def _orfaos_parecer(s: Session) -> list[Achado]:
    achados: list[Achado] = []
    for parecer in s.execute(select(ParecerTecnico)).scalars():
        if parecer.situacao == "RESERVADO":
            continue
        if parecer.processo_id is None:
            achados.append(
                Achado(
                    referencia=f"{parecer.numero}/{parecer.ano}",
                    detalhe="parecer sem processo vinculado",
                    parecer_id=parecer.id,
                )
            )
        if parecer.laudo_id is None and parecer.situacao != "RASCUNHO":
            achados.append(
                Achado(
                    referencia=f"{parecer.numero}/{parecer.ano}",
                    detalhe="parecer sem laudo — não há peça que caracterize a exposição",
                    parecer_id=parecer.id,
                )
            )
    return achados


def _conflitos_numeracao(s: Session) -> list[Achado]:
    achados: list[Achado] = []
    pareceres = {
        (p.numero, p.ano): p for p in s.execute(select(ParecerTecnico)).scalars()
    }
    for cartao in s.execute(
        select(StgTrelloCartao).where(StgTrelloCartao.parecer_candidato.isnot(None))
    ).scalars():
        candidato = cartao.parecer_candidato or parecer_candidato(cartao.nome)
        if not candidato or "/" not in candidato:
            continue
        numero, ano = candidato.split("/")
        parecer = pareceres.get((int(numero), int(ano)))
        if parecer is None:
            achados.append(
                Achado(
                    referencia=f"trello:{cartao.card_id}",
                    detalhe=(
                        f"o título '{cartao.nome}' aponta o parecer {candidato}, "
                        "que não existe na planilha"
                    ),
                )
            )
            continue
        nome_no_cartao = (cartao.nome or "").upper()
        nome_no_parecer = (parecer.servidor.nome if parecer.servidor else "").upper()
        if nome_no_parecer and nome_no_parecer.split()[0] not in nome_no_cartao:
            achados.append(
                Achado(
                    referencia=f"{candidato}",
                    detalhe=(
                        f"o cartão '{cartao.nome}' aponta o parecer {candidato}, "
                        f"mas na planilha esse número é de '{parecer.servidor.nome}'"
                    ),
                    parecer_id=parecer.id,
                )
            )
    for parecer in s.execute(
        select(ParecerTecnico).where(ParecerTecnico.situacao == "RESERVADO")
    ).scalars():
        achados.append(
            Achado(
                referencia=f"{parecer.numero}/{parecer.ano}",
                detalhe=(
                    "número reservado e nunca preenchido — conferir se não existe "
                    "parecer assinado com ele"
                ),
                parecer_id=parecer.id,
            )
        )
    return achados


def _nups_dv_invalido(s: Session) -> list[Achado]:
    achados: list[Achado] = []
    for processo in s.execute(select(Processo)).scalars():
        resultado = servico_nup.validar(processo.nup)
        if not resultado.dv_ok:
            achados.append(
                Achado(
                    referencia=processo.nup,
                    detalhe=(
                        "dígito verificador não confere"
                        + (" (dispensa registrada)" if processo.nup_dv_dispensado else "")
                    ),
                    processo_id=processo.id,
                )
            )
    return achados


def _divergencias_marco(s: Session) -> list[Achado]:
    achados: list[Achado] = []
    for parecer in s.execute(select(ParecerTecnico)).scalars():
        if parecer.data_marco_inicial is None or parecer.tipo_marco is None:
            continue
        if parecer.tipo_marco.codigo != "PORTARIA_LOCALIZACAO":
            if not parecer.justificativa_marco:
                achados.append(
                    Achado(
                        referencia=f"{parecer.numero}/{parecer.ano}",
                        detalhe=(
                            f"marco '{parecer.tipo_marco.rotulo}' em "
                            f"{datas_br.numerica(parecer.data_marco_inicial)} sem justificativa"
                        ),
                        parecer_id=parecer.id,
                    )
                )
            continue
        if parecer.portaria is None:
            achados.append(
                Achado(
                    referencia=f"{parecer.numero}/{parecer.ano}",
                    detalhe="marco é a portaria de localização, mas não há portaria vinculada",
                    parecer_id=parecer.id,
                )
            )
        elif parecer.data_marco_inicial != parecer.portaria.data_publicacao:
            achados.append(
                Achado(
                    referencia=f"{parecer.numero}/{parecer.ano}",
                    detalhe=(
                        f"marco {datas_br.numerica(parecer.data_marco_inicial)} × portaria "
                        f"{datas_br.numerica(parecer.portaria.data_publicacao)}"
                    ),
                    parecer_id=parecer.id,
                )
            )
    return achados


def reconciliar(s: Session) -> Reconciliacao:
    return Reconciliacao(
        orfaos_processo=_orfaos_processo(s),
        orfaos_parecer=_orfaos_parecer(s),
        conflitos_numeracao=_conflitos_numeracao(s),
        nups_dv_invalido=_nups_dv_invalido(s),
        divergencias_marco=_divergencias_marco(s),
    )


def rejeitadas(s: Session, limite: int = 200) -> list[MigracaoRejeitada]:
    return list(
        s.execute(
            select(MigracaoRejeitada).order_by(MigracaoRejeitada.id.desc()).limit(limite)
        ).scalars()
    )

File: nuvem/src/servicos/reconciliacao.ts
/**
 * Fase 4 — os cinco relatórios de reconciliação da migração.
 * Porte de `app/servicos/reconciliacao.py`.
 *
 *   1. órfãos de processo    — cartões/linhas sem NUP
 *   2. órfãos de parecer     — pareceres sem processo, e processos que
 *                              deveriam ter parecer e não têm
 *   3. conflito de numeração — o mesmo número apontando para coisas diferentes
 *   4. NUPs com DV inválido  — digitados sem conferência
 *   5. divergência de marco  — marco inicial que não bate com a portaria
 *
 * Nenhum deles resolve nada sozinho: expõem para o Fabrício decidir.
 */
import { asc, desc, eq, isNotNull, like } from "drizzle-orm";
import type { Executor } from "../db/cliente.js";
import {
  migracao_rejeitada,
  parecer_tecnico,
  processo as tabela_processo,
  stg_planilha_parecer,
  stg_trello_cartao,
} from "../db/esquema/index.js";
import * as datas_br from "./datas_br.js";
import * as servico_nup from "./nup.js";
import { parecer_candidato } from "./importacao_trello.js";

export interface Achado {
  referencia: string;
  detalhe: string;
  processo_id: number | null;
  parecer_id: number | null;
}

function achado(referencia: string, detalhe: string, ids: { processo_id?: number; parecer_id?: number } = {}): Achado {
  return { referencia, detalhe, processo_id: ids.processo_id ?? null, parecer_id: ids.parecer_id ?? null };
}

export class Reconciliacao {
  orfaos_processo: Achado[] = [];
  orfaos_parecer: Achado[] = [];
  conflitos_numeracao: Achado[] = [];
  nups_dv_invalido: Achado[] = [];
  divergencias_marco: Achado[] = [];

  get total(): number {
    return (
      this.orfaos_processo.length +
      this.orfaos_parecer.length +
      this.conflitos_numeracao.length +
      this.nups_dv_invalido.length +
      this.divergencias_marco.length
    );
  }

  como_secoes(): [string, string, Achado[]][] {
    return [
      [
        "Órfãos de processo",
        "cartões e linhas sem NUP — precisam do número real para virar processo",
        this.orfaos_processo,
      ],
      ["Órfãos de parecer", "pareceres sem processo vinculado, e o contrário", this.orfaos_parecer],
      [
        "Conflito de numeração",
        "o mesmo número apontando para coisas diferentes — não resolver automaticamente, decidir caso a caso",
        this.conflitos_numeracao,
      ],
      ["NUPs com dígito verificador inválido", "entraram com a flag de dispensa; confirmar no SEI", this.nups_dv_invalido],
      [
        "Divergência de marco inicial",
        "a data do marco não bate com a publicação da portaria (RN-05)",
        this.divergencias_marco,
      ],
    ];
  }
}

async function _orfaos_processo(tx: Executor): Promise<Achado[]> {
  const achados: Achado[] = [];
  for (const cartao of await tx.select().from(stg_trello_cartao).orderBy(asc(stg_trello_cartao.id))) {
    if (!servico_nup.extrair(cartao.descricao).length) {
      achados.push(achado(`trello:${cartao.card_id}`, `cartão '${cartao.nome}' na lista '${cartao.lista}' sem NUP`));
    }
  }
  for (const linha of await tx.select().from(stg_planilha_parecer).orderBy(asc(stg_planilha_parecer.id))) {
    if (linha.col_b_numero_parecer && !linha.col_k_numero_processo) {
      achados.push(
        achado(
          `${linha.arquivo}:${linha.linha_origem}`,
          `parecer ${linha.col_b_numero_parecer}/${linha.col_d_ano || "?"} sem nº de processo na planilha`,
        ),
      );
    }
  }
  // NUPs sintéticos criados na migração do Trello (ano 1900)
  for (const processo of await tx
    .select()
    .from(tabela_processo)
    .where(like(tabela_processo.nup, "23086.%/1900-%"))
    .orderBy(asc(tabela_processo.id))) {
    achados.push(
      achado(processo.nup, "NUP sintético gerado na migração — substituir pelo número real", { processo_id: processo.id }),
    );
  }
  return achados;
}

async function _orfaos_parecer(tx: Executor): Promise<Achado[]> {
  const achados: Achado[] = [];
  for (const parecer of await tx.select().from(parecer_tecnico).orderBy(asc(parecer_tecnico.id))) {
    if (parecer.situacao === "RESERVADO") continue;
    const ref = `${parecer.numero}/${parecer.ano}`;
    if (parecer.processo_id === null) achados.push(achado(ref, "parecer sem processo vinculado", { parecer_id: parecer.id }));
    if (parecer.laudo_id === null && parecer.situacao !== "RASCUNHO") {
      achados.push(
        achado(ref, "parecer sem laudo — não há peça que caracterize a exposição", { parecer_id: parecer.id }),
      );
    }
  }
  return achados;
}

async function _conflitos_numeracao(tx: Executor): Promise<Achado[]> {
  const achados: Achado[] = [];
  const pareceres = new Map<string, { id: number; servidor: { nome: string } | null }>();
  for (const p of await tx.query.parecer_tecnico.findMany({
    with: { servidor: true },
    orderBy: [asc(parecer_tecnico.id)],
  })) {
    pareceres.set(`${p.numero}/${p.ano}`, p);
  }
  for (const cartao of await tx
    .select()
    .from(stg_trello_cartao)
    .where(isNotNull(stg_trello_cartao.parecer_candidato))
    .orderBy(asc(stg_trello_cartao.id))) {
    const candidato = cartao.parecer_candidato || parecer_candidato(cartao.nome);
    if (!candidato || !candidato.includes("/")) continue;
    const [numero, ano] = candidato.split("/");
    const parecer = pareceres.get(`${Number(numero)}/${Number(ano)}`);
    if (!parecer) {
      achados.push(
        achado(
          `trello:${cartao.card_id}`,
          `o título '${cartao.nome}' aponta o parecer ${candidato}, que não existe na planilha`,
        ),
      );
      continue;
    }
    const nome_no_cartao = (cartao.nome || "").toUpperCase();
    const nome_no_parecer = (parecer.servidor ? parecer.servidor.nome : "").toUpperCase();
    // `.split()[0]` do Python: a primeira palavra, cortando em qualquer branco
    const primeiro = nome_no_parecer.trim().split(/\s+/)[0] ?? "";
    if (nome_no_parecer && !nome_no_cartao.includes(primeiro)) {
      achados.push(
        achado(
          candidato,
          `o cartão '${cartao.nome}' aponta o parecer ${candidato}, ` +
            `mas na planilha esse número é de '${parecer.servidor!.nome}'`,
          { parecer_id: parecer.id },
        ),
      );
    }
  }
  for (const parecer of await tx
    .select()
    .from(parecer_tecnico)
    .where(eq(parecer_tecnico.situacao, "RESERVADO"))
    .orderBy(asc(parecer_tecnico.id))) {
    achados.push(
      achado(
        `${parecer.numero}/${parecer.ano}`,
        "número reservado e nunca preenchido — conferir se não existe parecer assinado com ele",
        { parecer_id: parecer.id },
      ),
    );
  }
  return achados;
}

async function _nups_dv_invalido(tx: Executor): Promise<Achado[]> {
  const achados: Achado[] = [];
  for (const processo of await tx.select().from(tabela_processo).orderBy(asc(tabela_processo.id))) {
    const resultado = servico_nup.validar(processo.nup);
    if (!resultado.dv_ok) {
      achados.push(
        achado(
          processo.nup,
          "dígito verificador não confere" + (processo.nup_dv_dispensado ? " (dispensa registrada)" : ""),
          { processo_id: processo.id },
        ),
      );
    }
  }
  return achados;
}

async function _divergencias_marco(tx: Executor): Promise<Achado[]> {
  const achados: Achado[] = [];
  for (const parecer of await tx.query.parecer_tecnico.findMany({
    with: { tipo_marco: true, portaria: true },
    orderBy: [asc(parecer_tecnico.id)],
  })) {
    if (parecer.data_marco_inicial === null || parecer.tipo_marco === null) continue;
    const ref = `${parecer.numero}/${parecer.ano}`;
    if (parecer.tipo_marco.codigo !== "PORTARIA_LOCALIZACAO") {
      if (!parecer.justificativa_marco) {
        achados.push(
          achado(
            ref,
            `marco '${parecer.tipo_marco.rotulo}' em ${datas_br.numerica(parecer.data_marco_inicial)} sem justificativa`,
            { parecer_id: parecer.id },
          ),
        );
      }
      continue;
    }
    if (parecer.portaria === null) {
      achados.push(
        achado(ref, "marco é a portaria de localização, mas não há portaria vinculada", { parecer_id: parecer.id }),
      );
    } else if (parecer.data_marco_inicial !== parecer.portaria.data_publicacao) {
      achados.push(
        achado(
          ref,
          `marco ${datas_br.numerica(parecer.data_marco_inicial)} × portaria ` +
            `${datas_br.numerica(parecer.portaria.data_publicacao)}`,
          { parecer_id: parecer.id },
        ),
      );
    }
  }
  return achados;
}

export async function reconciliar(tx: Executor): Promise<Reconciliacao> {
  const r = new Reconciliacao();
  r.orfaos_processo = await _orfaos_processo(tx);
  r.orfaos_parecer = await _orfaos_parecer(tx);
  r.conflitos_numeracao = await _conflitos_numeracao(tx);
  r.nups_dv_invalido = await _nups_dv_invalido(tx);
  r.divergencias_marco = await _divergencias_marco(tx);
  return r;
}

export async function rejeitadas(tx: Executor, limite = 200) {
  return tx.select().from(migracao_rejeitada).orderBy(desc(migracao_rejeitada.id)).limit(limite);
}
